mirrord/layer/src/socket/ops.rs

```diff
@@ -413,6 +417,53 @@ pub(super) fn listen(sockfd: RawFd, backlog: c_int) -> Detour<i32> {
     }
 }
 
+fn non_blocking_connect_watcher(
```

Comment:
I'm trying to understand what's going on with this, and wondering if there isn't a more straightforward solution to the issue.

I wonder: is there a reason why we can't have a secondary thread in the layer, that loops checking for some sort of `BackgroundMessage`, and drop this necessity of having all of our messages be request/responses? Wouldn't something like this simplify all of this?

Elaborating on what I'm thinking is something like: layer sends the usual connect request, intproxy responds to it immediately (either with a success or `EINPROGRESS`), user gets the result of this operation, and we put the socket in some `InProgress(remote_addr/whatever)` state. On this secondary thread, at some point we receive `Connected(id, stuff)`, and change the socket to be in a connected state.

I think something like this would be sort of a `poll` implementation?

Reply:
If my idea is dumb, then please add docs for this function.